From 832e4fae367c621c24a7b6ff65728dc06ebd7fc8 Mon Sep 17 00:00:00 2001
From: Emmanuel Blot <emmanuel.blot@free.fr>
Date: Thu, 6 Aug 2026 11:09:35 +0200
Subject: [PATCH] tests/functional/arm: clemente: add BMC boot test

Boot the OpenBMC flash image, then drive the two OCP NIC TMP421
sensors through QOM and confirm the readings propagate to the guest
hwmon interface.

Signed-off-by: Emmanuel Blot <emmanuel.blot@free.fr>
---
 tests/functional/arm/meson.build             |  2 +
 tests/functional/arm/test_aspeed_clemente.py | 64 ++++++++++++++++++++
 2 files changed, 66 insertions(+)
 create mode 100644 tests/functional/arm/test_aspeed_clemente.py

diff --git a/tests/functional/arm/meson.build b/tests/functional/arm/meson.build
index 81c67e3733..e82274b723 100644
--- a/tests/functional/arm/meson.build
+++ b/tests/functional/arm/meson.build
@@ -15,6 +15,7 @@ test_arm_timeouts = {
   'aspeed_anacapa' : 480,
   'aspeed_bletchley' : 480,
   'aspeed_catalina' : 480,
+  'aspeed_clemente' : 480,
   'aspeed_gb200nvl_bmc' : 480,
   'aspeed_minerva' : 480,
   'aspeed_rainier' : 480,
@@ -53,6 +54,7 @@ tests_arm_system_thorough = [
   'aspeed_anacapa',
   'aspeed_bletchley',
   'aspeed_catalina',
+  'aspeed_clemente',
   'aspeed_gb200nvl_bmc',
   'aspeed_minerva',
   'aspeed_rainier',
diff --git a/tests/functional/arm/test_aspeed_clemente.py b/tests/functional/arm/test_aspeed_clemente.py
new file mode 100644
index 0000000000..d30e3662cc
--- /dev/null
+++ b/tests/functional/arm/test_aspeed_clemente.py
@@ -0,0 +1,64 @@
+#!/usr/bin/env python3
+#
+# Functional test that boots the ASPEED machines
+#
+# Copyright (c) 2026 Meta Platforms, Inc. and affiliates.
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+from qemu_test import Asset
+from qemu_test import exec_command_and_wait_for_pattern
+from aspeed import AspeedTest
+
+
+class ClementeMachine(AspeedTest):
+
+    # TODO: the boot image currently lives on a personal fork; move it under
+    # legoater/qemu-aspeed-boot before upstreaming.
+    ASSET_CLEMENTE_FLASH = Asset(
+        'https://github.com/eblot/qemu-aspeed-boot/raw/refs/heads/'
+            'clemente-bmc/images/clemente-bmc/openbmc-20260902025058/'
+            'obmc-phosphor-image-clemente-20260902025058.static.mtd.xz',
+        'ddd9ecd9793892ba0834ac466d66835353fd8125de0cfedf756d3fe8a7e23e0d')
+
+    # The OCP NIC TMP421 sensors sit directly on i2c10 and i2c15 at address
+    # 0x1f; on each bus the sensor is the first slave instantiated (child[1],
+    # after the controller's own child[0]). They enumerate in the guest as
+    # devices 10-001f and 15-001f.
+    NIC0_QOM_PATH = "/machine/soc/i2c/bus[10]/aspeed.i2c.bus.10/child[1]"
+    NIC0_HWMON = "/sys/bus/i2c/devices/10-001f/hwmon/hwmon*"
+    NIC1_QOM_PATH = "/machine/soc/i2c/bus[15]/aspeed.i2c.bus.15/child[1]"
+    NIC1_HWMON = "/sys/bus/i2c/devices/15-001f/hwmon/hwmon*"
+    PROMPT = "root@clemente:~#"
+
+    def test_arm_ast2600_clemente_openbmc(self):
+        image_path = self.uncompress(self.ASSET_CLEMENTE_FLASH)
+
+        self.do_test_arm_aspeed_openbmc('clemente-bmc', image=image_path,
+                                        uboot='2019.04', cpu_id='0xf00',
+                                        soc='AST2600 rev A3')
+
+        exec_command_and_wait_for_pattern(self, 'root', 'Password:')
+        exec_command_and_wait_for_pattern(self, '0penBmc', self.PROMPT)
+
+        # temperature0 is the local channel (hwmon temp1_input), temperature1
+        # the first remote channel (temp2_input), both in millidegrees C. A
+        # set point on the TMP421's 1/16 C grid survives the round trip; one
+        # off the grid truncates down to it.
+        for qom, hwmon in ((self.NIC0_QOM_PATH, self.NIC0_HWMON),
+                           (self.NIC1_QOM_PATH, self.NIC1_HWMON)):
+            self.assertIn(b"tmp421", self.read_hwmon(hwmon, "name"))
+
+            for (local_mc, local_rb), (remote_mc, remote_rb) in \
+                    (((30000, 30000), (55000, 55000)),
+                     ((25020, 25000), (45020, 45000))):
+                self.vm.cmd("qom-set", path=qom,
+                            property="temperature0", value=local_mc)
+                self.vm.cmd("qom-set", path=qom,
+                            property="temperature1", value=remote_mc)
+                self.wait_hwmon_value(hwmon, "temp1_input", local_rb)
+                self.wait_hwmon_value(hwmon, "temp2_input", remote_rb)
+
+
+if __name__ == '__main__':
+    AspeedTest.main()
-- 
2.53.0-Meta

